sdks/python/apache_beam/metrics/internal.py

```diff
@@ -0,0 +1,116 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Internal classes for Metrics API.
+
+The classes in this file keep shared state, and organize metrics information.
+
+Available classes:
+- MetricsEnvironment - Keeps track of MetricsContainer and other metrics
+    information for every single execution working thread.
+- MetricsContainer - Holds the metrics of a single step and a single
+    unit-of-commit (bundle).
+"""
+from collections import defaultdict
+import threading
+
+from apache_beam.metrics.base import MetricKey, MetricUpdates
+from apache_beam.metrics.cells import CounterCell, DistributionCell
+
+
+class MetricsEnvironment(object):
+  """ Holds the MetricsContainer for every thread and other metric information.
+
+  This class is not meant to be instantiated, instead being used as to keep
+  track of global state.
+  """
+  METRICS_SUPPORTED = False
+  _METRICS_SUPPORTED_LOCK = threading.Lock()
+
+  PER_THREAD = threading.local()
+
+  @classmethod
+  def set_metrics_supported(cls, supported):
```

Comment:
You may want to look at the recent java changes. The MetricsEnvironment now supports scoped updates, such as:

```
with MetricsEnvironment.set_current_container(container):
  ...
```

That automatically resets it to the previous container. May be useful.